indexers/keyvalue/LevelDBIndexer/tests/test_leveldbindexer.py

```diff
@@ -19,43 +20,169 @@ def rm_files(file_paths):
+            _, documents, _ = get_documents(ids)
+            validate_positive_results(keys, documents, searcher)
+        if negative_results:
+            validate_negative_results(negative_results, searcher)
 
-    rm_files([save_abspath, index_abspath])
 
+def run_crud_test_for_scenario(indexer, actions, results, no_results):
+    exception = None
+    try:
+        with indexer:
+            indexer.touch()
+            indexer.save()
+        save_abspath = indexer.save_abspath
+        index_abspath = indexer.index_abspath
+        assert os.path.exists(index_abspath)
+        assert os.path.exists(save_abspath)
+        apply_actions(save_abspath, index_abspath, actions)
+        validate_results(save_abspath, results, no_results)
+    except Exception as e:
+        exception = e
+    finally:
+        rm_files([save_abspath, index_abspath])
+        if exception is not None:
+            raise exception
 
-def test_add_query():
-    indexer = LevelDBIndexer(level='doc', index_filename='leveldb.db')
-    run_test(indexer)
 
+def run_crud_test(actions, results, no_results):
+    # test construction from code
+    indexer = LevelDBIndexer(level='doc', index_filename='leveldb.db')
+    run_crud_test_for_scenario(indexer, actions, results, no_results)
 
-def test_load_yaml():
+    # test construction from yaml
     from jina.executors import BaseExecutor
     indexer = BaseExecutor.load_config(os.path.join(cur_dir, 'yaml/test-leveldb.yml'))
-    run_test(indexer)
+    run_crud_test_for_scenario(indexer, actions, results, no_results)
+
+
+# action = (method, key, document_id)
+
+@pytest.mark.parametrize('actions, results, no_results, exception', [
+    # basic add
+    (
+            [('add', {0: 0, 1: 1}), ('add', {3: 3})],
+            {0: 0, 1: 1, 3: 3},
+            [2],
+            None
+    ),
+    # add existing key updates the content
+    (
+            [('add', {0: 0, 1: 1}), ('add', {0: 3})],
+            {0: 3, 1: 1},
+            [2],
+            None
+    ),
+    # update existing key
+    (
+            [('add', {1: 1, 2: 2}), ('update', {1: 4})],
+            {1: 4, 2: 2},
+            [0, 3],
+            None
+    ),
+    # update non existing key
+    (
+            [('update', {1: 4})],
+            {},
+            [0, 3],
+            KeyError
+    ),
+    # update non existing key and existing key
+    (
+            [('add', {1: 1, 2: 2}), ('update', {1: 4, 3: 4})],
+            {},
+            [0, 3],
+            KeyError
+    ),
+    # updating two keys to have the same value is allowed
+    (
+            [('add', {1: 1, 2: 2}),('update', {1: 2})],
+            {},
+            [0, 3],
+            None
+    ),
+    # longer chain of actions
+    (
+            [('add', {0: 0, 1: 1}), ('delete', {1: 1}), ('add', {3: 3, 9:4, 2:4}), ('delete', {0: 0, 2: 4})],
+            {9: 4, 3: 3},
+            [0, 1, 4],
+            None
+    ),
+])
+def test_crud(actions, results, no_results, exception):
```

Comment:
This is nota very pytesty style. Verter separate the test on diferent named test for better readability and failure understanding.

Reply:
sure I changed it to single tests.